Add bidi support

ahangarha opened the pull request:
Writing in LTR based forums is always a headache for people who use RTL languages. To add LTR/RTL support, we need to use auto direction (`dir="auto"`) to tags which can potentially include either of RTL or LTR text.

My commits handled this in two steps:
1. add `dir="auto"` to relevant tags in `parser.php` file
2. remove unnecessary `text-align: left` from styles` css.

Just notice that I have not been able to text the changes since I faced with difficulties for running FluxBB locally. But I don't think there would be any issue.

In order to verify changes, copy some RTL text in a post. You should see the paragraph starts with RTL character should go to the right and becomes RTL and paragraph (or heading or quotation) starting with English letters should remain as normal.

ahangarha committed:
add auto direction attr to elemets

ahangarha committed:
remove unneeded text-align from Air style

ahangarha committed:
remove unneeded text-align from Earth style

ahangarha committed:
remove unneeded text-align from Fire style

FTABoy commented:
i don't know why but sadly it broke posting. when i try to send a new post (or preview) i get this error:
 Catchable fatal error: Object of class Closure could not be converted to string in D:\..........\include\functions.php on line 2059

ahangarha commented:
I just could successfully run fluxbb locally. I faced with no issue in posting. this is an example of post with both LTR and RTL paragraphs:
![commit result](https://user-images.githubusercontent.com/11241315/58731809-0845b600-8405-11e9-8297-a97f029eda3c.png)

ahangarha committed:
add bidi support to textareas

franzliedke commented:
Thanks for this change, this is awesome!

I would appreciate if you could test these changes in detail, though - if you need help with getting your local copy running, feel free to open a topic in our [support forums](https://fluxbb.org/forums).

FTABoy commented:
> 
> 
> Thanks for this change, this is awesome!
> 
> I would appreciate if you could test these changes in detail, though - if you need help with getting your local copy running, feel free to open a topic in our [support forums](https://fluxbb.org/forums).

hi, as my friend Mr Ahangarha has mentioned he successfully managed to run his local copy of fluxbb and tested everything. also i will try to test this again. thanks

ahangarha commented:
@franzliedke sure

ahangarha commented:
@franzliedke I though that link has some instruction for a series of predefined tests.

I have run the app on my local machine and it worked perfectly as expected. I sent screenshot as well. My env:
- Kubuntu 19.04
- PHP 7.2.17-0ubuntu0.19.04.1 (cli)
- mysql  Ver 14.14 Distrib 5.7.26

But @FTABoy has issue in posting. As far as I know, he runs on PHP 5.6 and gets error while posting. He himself might explain in detail.

nahos2585 reviewed (changes requested):
also bbcodes like [list][*] tag should be considered to support bidi (RTL).
every thing else is working perfect

FTABoy commented:
> 
> 
> But @FTABoy has issue in posting. As far as I know, he runs on PHP 5.6 and gets error while posting. He himself might explain in detail.
i tried again with a fresh clone of your @ahangarha repository bidi branch , posting and previews are working with out any problem. but i get an other different error on top of the page :
 Strict Standards: Only variables should be passed by reference in D:\AppServ\www\127.0.0.1\fluxbb\include\functions.php on line 582
![error](https://user-images.githubusercontent.com/25687276/58807909-3bb75900-862e-11e9-8c6f-9b77f1e88cd4.jpg)

ahangarha commented:
I think your issue is not related to my changes. I think it should be a general issue with your version of PHP or MySql. Please check if you have this issue on master branch or not. if it remains, I suggest you open a separated issue for it.

ahangarha commented:
> also bbcodes like [list][*] tag should be considered to support bidi (RTL).
> every thing else is working perfect

I couldn't get any progress on this specific case! Can anyone help on this issue?

ahangarha committed:
Fix bidi for UL and OL

To let UL and OL detect the currect direction of the inner text, it is
needed to put the text directly inside <li> without pragraph tag around
it.

ahangarha committed:
Remove dir=auto from quotations

Adding dir=auto was over doing since the content of the quatation itself
contains dir=auto. So it is undone in this commit.

ahangarha commented:
I finally figured out the problem and solved it in commit [31fbf67](https://github.com/fluxbb/fluxbb/pull/227/commits/31fbf67044e664fbd587b0f740d93d28d341770e).

As per my understanding, this PR can be merged. I have not seen any malfunction.

ahangarha commented:
Is there any problem for merging this PR?

franzliedke commented on include/parser.php:
```diff
@@ -749,17 +749,17 @@ function ($match) { return handle_list_tag($match[2], $match[1]); },
 	}
 
-	$content = preg_replace('#\s*\[\*\](.*?)\[/\*\]\s*#s', '<li><p>$1</p></li>', pun_trim($content));
+	$content = preg_replace('#\s*\[\*\](.*?)\[/\*\]\s*#s', '<li>$1</li>', pun_trim($content));
```
Why is the paragraph tag being removed here? That seems like an unwanted change to the current markup?

ahangarha replied:
It was made to have correct impact. But I think I need to revise this!

Still you can merge it without `p` which is the right thing to do. Later I will share some workaround if needed

ahangarha replied:
Having `p` in the `li` makes issue. If it is necessary to have paragraph in list, let me know so I make required changes.

franzliedke commented on style/Earth.css:
```diff
@@ -731,7 +731,6 @@ MAIN POSTS
 	float: right;
 	margin-right: -218px;
 	position: relative;
-	text-align: left;
```
So these rules are superfluous in the default, left-to-right layout?

ahangarha replied:
There shouldn't be any absolute left or right unless you are sure that regardless of the text direction, the text should remain on left or right.

If you need to mention the alignment, use start|end instead of left|right. I removed it as I found it unnecessary.

franzliedke commented on viewtopic.php:
```diff
@@ -456,7 +456,7 @@
 
 ?>
-<textarea name="req_message" rows="7" cols="75" tabindex="<?php echo $cur_index++ ?>"></textarea></label>
+<textarea dir="auto" name="req_message" rows="7" cols="75" tabindex="<?php echo $cur_index++ ?>"></textarea></label>
```
Out of interest: why are changes like these only made to `textarea` tags, but no `input` elements?

ahangarha replied:
If we want to solve this problem everywhere, then it should be applied many places. My changes addressed the most serious issues.

ahangarha replied:
To complete my last comment, here I added dir=auto to textarea to get the closest possible experience during edit with the final html result.

If we want to change inputs, we need to make sure the html output also supports bidi

franzliedke reviewed (commented):
Thanks for your patience! I will definitely merge this in time for the next release, but I've been terribly busy...

I collected a few questions inline, to better understand all the changes.

ahangarha commented:
Tnx. Another forum needs this to be implemented and are waiting for it.
They prefer to use master and not using cherry pick method to get changes.

Any possible schedule for release? Any help you need?

در تاریخ چهارشنبه ۲ اکتبر ۲۰۱۹،‏ ۳:۵۴ Franz Liedke <notifications@github.com>
نوشت:

> *@franzliedke* commented on this pull request.
>
> Thanks for your patience! I will definitely merge this in time for the
> next release, but I've been terribly busy...
>
> I collected a few questions inline, to better understand all the changes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/fluxbb/fluxbb/pull/227?email_source=notifications&email_token=ACVYOYZ2SNIVYSUCM6A4Q5TQMPS2HA5CNFSM4HR46QQKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCGSALAQ#pullrequestreview-295961986>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACVYOY5B77EXSJYHXA3XIO3QMPS2HANCNFSM4HR46QQA>
> .
>

franzliedke commented:
I am about to head out on vacation for two weeks. After that, I will gladly try to package up the unfinished release.

> Any help you need?

If you're interested, you can try to tackle these two issues:
- https://fluxbb.org/development/core/tickets/1092/
- https://fluxbb.org/development/core/tickets/1123/

ahangarha commented:
Hi

Any update?

ahangarha commented:
any update?

natrius commented:
@franzliedke Any chance to get an update on this anyway? Solving this PR has nothing to do with a release, to be honest.

ahangarha commented:
Regarding update, I mean status of the PR. If you are waiting for some modification from my side, please let me know. If not, This can be released as a patch for a subversion. I know there are forums waiting for such fixes.

ahangarha commented:
Any update regarding this PR?

ahangarha commented:
Any update?
